### PR TITLE
Data structure wip

### PR DESCRIPTION
**Pull request by lukechurch**
@as2388 Any initial comments?

### CODE REVIEW

File: data_tools/schema.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            other.version == self.version & \
+            other.documentation == self.documentation & \
+            other.codes == self.codes
+    
+    def __neq__(self, other):
+        return not self.__eq__(other)
+
+
+    @staticmethod
+    def from_map(data):
```
Comment (as2388) [inline data_tools/schema.py]: Wonder if `from_dict` (or even `load_dict`) is more Pythonic

File: data_tools/schema.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import validators
+
+class Schema(object):
+    scheme_id = None
+    name = None
+    version = None
+    codes = []
+    documentation = {}
```
Comment (as2388) [inline data_tools/schema.py]: I always get confused over whether `{}` is a dict or a set

File: data_tools/schema.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import validators
+
+class Schema(object):
```
Comment (as2388) [inline data_tools/schema.py]: `Schema` -> `Scheme`

**Comment by as2388** [comment]
@lukechurch also this doesn't load directly from a file. I feel like there should be library code somewhere which does that.

File: data_tools/schema.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return code
+    
+    def to_map(self):
+        ret = {}
+        ret["CodeID"] = validators.validate_string(self.code_id, "CodeID")
+        ret["DisplayText"] = validators.validate_string(self.display_text, "DisplayText")
+        ret["Shortcut"] = validators.validate_string(self.shortcut, "Shortcut")
+        ret["NumericValue"] = validators.validate_int(self.numeric_value, "NumericValue")
+        ret["VisibleInCoda"] = validators.validate_bool(self.visible_in_coda, "VisibleInCoda")
+        ret["Color"] = validators.validate_string(self.color, "Color")
```
[the remark this answers is not in the document]
Comment (lukechurch) [reply data_tools/schema.py]: done

File: data_tools/schema.py (hunk 1 above)
Comment (lukechurch) [reply data_tools/schema.py]: Done per discussion

File: data_tools/schema.py (hunk 2 above)
Comment (lukechurch) [reply data_tools/schema.py]: done

File: data_tools/schema.py (hunk 3 above)
Comment (lukechurch) [reply data_tools/schema.py]: done

**Comment by lukechurch** [comment]
> @lukechurch also this doesn't load directly from a file. I feel like there should be library code somewhere which does that.

As discussed - I agree, needs to exist in a slightly different place

**Comment by lukechurch** [comment]
Thanks @lukechurch

File: data_tools/scheme.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    numeric_value = -42
+    visible_in_coda = True
+    color = None
+
+    @staticmethod
+    def from_firebase_map(data):
+        code = Code()
+        code.code_id = validators.validate_string(data["CodeID"], "CodeID")
+        code.display_text = validators.validate_string(data["DisplayText"], "DisplayText")
+        if "ShortCut" in data.keys():
```
Comment (marianamarasoiu) [inline data_tools/scheme.py]: typo: I think it should be "Shortcut" (lowercase c)

Comment (lukechurch) [reply data_tools/scheme.py]: Thanks!

**Pull request merged**